### PR TITLE
Kaldi datalayer

### PR DESCRIPTION
**Pull request by redoctopus**
Add Kaldi compatibility datalayer and dataset to nemo_asr. Reads in MFCC features, transcripts, and (optionally) durations from Kaldi-formatted data. Uses kaldi-io.

### CODE REVIEW

**Comment by blisc** [comment]
Why kaldi_io instead of torchaudio?

**Comment by redoctopus** [comment]
It turns out that torchaudio.kaldi_io is just a wrapper around the kaldi_io library with a bit extra that we don't use. We really only need the one function from kaldi_io.

**Comment by blisc** [comment]
Let's rework Manifests again.
The tokenize_transcript function of ManifestBase should be a static function that accepts labels_map, unk_index, and blank_index instead of using self.

We have two options to rework ManifestEN. We can either make normalize_text a static function as well, or we define it outside of ManifestEN and have ManifestEN.normalize_text call this new function.

**Comment by redoctopus** [comment]
Did a bit of refactoring based on our discussion yesterday

File: collections/nemo_asr/nemo_asr/parts/manifest.py [hunk 1]
```diff
@@ -157,18 +161,18 @@ def normalize_text(self, text, labels):
 
         # Turn all other punctuation to whitespace
         table = str.maketrans(punctuation, " " * len(punctuation))
-        text = self.english_string_normalization(
+        text = ManifestEN.english_string_normalization(
```
Comment (blisc) [inline collections/nemo_asr/nemo_asr/parts/manifest.py]: Why not self.english_string_normalization()?

Comment (redoctopus) [reply collections/nemo_asr/nemo_asr/parts/manifest.py]: `normalize_text` is static (`name 'self' is not defined).

Maybe this is a good excuse to decouple the transcript processing functions from the Manifest class? e.g. transcript_tokenize_basic, text_normalization_basic, transcript_tokenize_EN, text_normalization_EN

Comment (blisc) [reply collections/nemo_asr/nemo_asr/parts/manifest.py]: I've been thinking about the onion discussion lately. There really is no need for the Manifest class at all. All it does it text cleaning and tokenization, which could all be helper functions. But I think that's beyond the scope of this PR.

If english_string_normalization is not used anywhere else, let's move it into ManifestEN.normalize_text()

File: collections/nemo_asr/nemo_asr/parts/manifest.py [hunk 2]
```diff
@@ -157,35 +161,10 @@ def normalize_text(self, text, labels):
...
-        """
         try:
-            text = clean_fn(s, table, punctuation_to_replace)
-            return text
+            text = clean_text(text, table, punctuation_to_replace)
         except BaseException:
             print("WARNING: Normalizing {} failed".format(s))
-            return None
+
+        return text
```
Comment (blisc) [inline collections/nemo_asr/nemo_asr/parts/manifest.py]: What happens if we catch BaseException? Would we crash because text is undefined?

Comment (redoctopus) [reply collections/nemo_asr/nemo_asr/parts/manifest.py]: `text` in that case would just be the original input (first argument).

Comment (blisc) [reply collections/nemo_asr/nemo_asr/parts/manifest.py]: The previous default here: https://github.com/NVIDIA/NeMo/blob/547a326369802f178e0be4da10ffea9ac0f30feb/collections/nemo_asr/nemo_asr/parts/manifest.py#L44
was to ignore that transcript if we couldn't normalize it, like NaNs in common voice. This is why we returned None instead of the unormalized string. Is there a reason to change the default behaviour or am I misunderstanding how the code works now?

Comment (redoctopus) [reply collections/nemo_asr/nemo_asr/parts/manifest.py]: Removed by accident, fixing now

File: collections/nemo_asr/nemo_asr/parts/manifest.py (hunk 1 above)
Comment (jonathan-cohen-nvidia) [reply collections/nemo_asr/nemo_asr/parts/manifest.py]: why wouldn't text normalization simply be another neural module?  That is much more in keeping with the NM design concept - we are defining a data flow graph of transformations.  Otherwise you are simply creating a separate non-generic framework for a specific data processing step.  That is counter to the overall system design.

Comment (blisc) [reply collections/nemo_asr/nemo_asr/parts/manifest.py]: There is no option nor support for any Neural Modules before a DataLayerNM. A DataLayerNM must be the first module in your call chain in the current design.
Of course you can just return unnormalized text from your dataloader but it is much easier to prune it prior to constructing your dataset than to do it on the fly during training.
